[TD] Enable Test Class granularity on heuristics

ZainRizvi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112162
* #112166
* __->__ #112161

Changes the heuristic framework to support multiple prioritizing individual classes within a test file.

Components of this included:
- Updating TestPrioritizations to accept individual test classes being prioritized. Previously, when a heuristic wanted to prioritize a test file it would pass in the test's name, now to prioritize a class within a test it uses the notation "test::classname"
- Changes are fully backwards compatible with existing heuristics
- Test sharding now supports sharding individual tests (for when they're prioritized)
- When a TestClass is prioritized, we pass the appropriate "-k" flags down to pytest

ZainRizvi committed:
TD: Enable Test Class granularity on heuristics

[ghstack-poisoned]

ZainRizvi committed:
Update on "[TD] Enable Test Class granularity on heuristics"


Changes the heuristic framework to support multiple prioritizing individual classes within a test file.

Components of this included:
- Updating TestPrioritizations to accept individual test classes being prioritized. Previously, when a heuristic wanted to prioritize a test file it would pass in the test's name, now to prioritize a class within a test it uses the notation "test::classname"
- Changes are fully backwards compatible with existing heuristics
- Test sharding now supports sharding individual tests (for when they're prioritized)
- When a TestClass is prioritized, we pass the appropriate "-k" flags down to pytest


[ghstack-poisoned]

ZainRizvi committed:
Update on "[TD] Enable Test Class granularity on heuristics"


Changes the heuristic framework to support multiple prioritizing individual classes within a test file.

Components of this included:
- Updating TestPrioritizations to accept individual test classes being prioritized. Previously, when a heuristic wanted to prioritize a test file it would pass in the test's name, now to prioritize a class within a test it uses the notation "test::classname"
- Changes are fully backwards compatible with existing heuristics
- Test sharding now supports sharding individual tests (for when they're prioritized)
- When a TestClass is prioritized, we pass the appropriate "-k" flags down to pytest


[ghstack-poisoned]

clee2000 commented on tools/testing/target_determination/heuristics/interface.py:
```diff
@@ -159,34 +254,37 @@ def _print_tests(label: str, tests: StringTuple) -> None:
+        for relevance_group, tests in self._traverse_priorities():
             _print_tests(f"{Relevance(relevance_group).name.title()} Relevance", tests)
 
     def _get_test_relevance_group(self, test_name: str) -> Relevance:
```
nit but probably specify that this is the highest relevance for any subset of this test, also probably for the below functions

ZainRizvi replied:
Good call

ZainRizvi commented on tools/testing/test_run.py:
```diff
@@ -0,0 +1,296 @@
+from typing import Iterable, List, Optional, Set, Tuple, Union
+
+
+class TestRun:
```
Yes, that assumption is definitely baked in. Good call that it should be explicitly stated.

ZainRizvi commented on tools/testing/test_run.py:
```diff
@@ -0,0 +1,296 @@
+        """
+        To subtract test runs means to run all the tests in the first run except for what the second run specifies.
+
+        It is currently an error if the subtraction will result in no tests being run.
```
Ty, I had realized it didn't have to be an error. Comment updated

ZainRizvi committed:
Update on "[TD] Enable Test Class granularity on heuristics"


Changes the heuristic framework to support multiple prioritizing individual classes within a test file.

Components of this included:
- Updating TestPrioritizations to accept individual test classes being prioritized. Previously, when a heuristic wanted to prioritize a test file it would pass in the test's name, now to prioritize a class within a test it uses the notation "test::classname"
- Changes are fully backwards compatible with existing heuristics
- Test sharding now supports sharding individual tests (for when they're prioritized)
- When a TestClass is prioritized, we pass the appropriate "-k" flags down to pytest


[ghstack-poisoned]

ZainRizvi commented:
@pytorchbot merge

ZainRizvi committed:
Update on "[TD] Enable Test Class granularity on heuristics"


Changes the heuristic framework to support multiple prioritizing individual classes within a test file.

Components of this included:
- Updating TestPrioritizations to accept individual test classes being prioritized. Previously, when a heuristic wanted to prioritize a test file it would pass in the test's name, now to prioritize a class within a test it uses the notation "test::classname"
- Changes are fully backwards compatible with existing heuristics
- Test sharding now supports sharding individual tests (for when they're prioritized)
- When a TestClass is prioritized, we pass the appropriate "-k" flags down to pytest


[ghstack-poisoned]

ZainRizvi commented:
@pytorchbot merge
